### ISSUE TITLE
Send and sniff?

### ISSUE THREAD

**Issue by throughthetrees** [ElectronicCats/flipper-rs485modbus#8]
So I've been able to test both the send and sniff components of the flipper modbus app on the flipper, but I have a question about sniffing versus sending.

I've been trying to use manual sender with 

Peripheral ID: 1
Function: 0x01
Start Address: 0
Quantity: 12

And I get back a time out

Baudrate: 4800
Data width: 8
Stop bits: 1
Parity: None
Response TimeOut: 500ms

![a473307023a39391651fc7ae7c10cc08fdf11a97-1](https://github.com/user-attachments/assets/d92d674f-ca29-4a2e-9027-e9d7e9d6f51f)

Which looks like a time out response. As in, it doesn't seem like I'm getting a response back from the device.

However, if I drop it into "Sniffer", and bridge the a and b lines with a RS-485 to USB device I have and send what I think is the same command from python, the sniffer picks up the packet and reads it correctly.

![7caa7104c05fd854f6b0bcb4e41ed6fbfd3beae2-1](https://github.com/user-attachments/assets/cb27235a-0545-4fd7-9d75-953d24cee5aa)

![d0c3dc83665a894684c3dfc579757333b3e6f97d-1](https://github.com/user-attachments/assets/bb2bb2fe-63c6-4998-ba46-77b0ccc97480)

-----PERIPHERAL-----
Read Holding Registers(03)
Peripheral: 1
bCount 20
->Reg0: 0
->Reg1: 527
->Reg2: 0
->Reg3: 0
->Reg4: 0
->Reg5: 0
->Reg6: 0
->Reg7: 0
->Reg8: 3072
->Reg9: 512
->Reg10: 0
-> Reg11: 0
CRC: 0x200B

So, if I had the "Sender" configured appropriately, I should get back the same response I get from the "Sniffer" ay? It does look like sniffer works correctly, but have I mis-configured the packet I'm sending to the device? It seems identical to the one I am sending via python.

So to test this, I put the "Sender" into the same config, but gave it a very long time out. I then send a packet from python. When I do this, I do get a response (as if the mod bus shield had sent the packet), and it looks like the right response.

So I think I've got the sender configured correctly but maybe it isn't actually sending the packet?

tldr: it seems like the send packet isn't working.

**Comment by throughthetrees** [comment]
The mystery deepens..

So using my voltmeter, I was able to test a few things. Looking at either or A or B to ground. I get a voltage response when I make a request from python via a USB to RS485.

Sniffing this packet allows me to use it via Buffer Sender, which I had not considered looking into previously. When I use this packet and send via Buffer Sender, I get a voltage response! Hooray! However, I am encountering errors "Length-Type MissMatch!!!" and "Check Response TimeOut!!!".

It seems like I'm getting some bytes back, but it doesn't look like they are being parsed correctly.


So:

Send via Buffer Send works. However, it doesn't seem to be parsing the result correctly?

Send Via Manual Send does not seem to work (No change in voltage when a packet is to be sent).

Will continue to investigate..

**Comment by xpeqex** [comment]
Hi @throughthetrees 

Thank you for reporting this and add detailed information about it. 

We want to get as much information as possible about this issue to work on it, so please share with us:
1. What Flipper firmware and version are you suing?
2. What version of the Modbus app are you using?
3. Please add a picture of each side of the Modbus Add-On

**Comment by xpeqex** [comment]
Thank you for the info.

So Flipper is registering function [0x03 which is READ HOLDING REGISTERS](https://github.com/ElectronicCats/flipper-rs485modbus/wiki#3-build-the-packet), the same as sent using the function client.read_holding_registers in your Python script.

So let's try something different:

- What happens if you create a new packet using Manual Sender in the Flipper app but now sending function 0x03?
- Could you please share with us the complete output on the Python side?
- What device are you using as a client/slave?
- Is there constant traffic on the bus? A server-client (master-slave) communication, other than the Flipper messages?

**Comment by throughthetrees** [comment]
So yes! It was my misunderstanding about what exactly is going on with read registers/ RS-485 commands.

Manual send is working!

However, it is still not registering the response correctly (in either manual or buffer send).


https://github.com/user-attachments/assets/87f3f4d3-f24d-42f5-aa20-2e5181d9e693

**Comment by xpeqex** [comment]
Ok, no worries. 

Looks like Flipper is interpreting the packet as a HUB request. Try setting the timeout to a larger value, it should not be that big but try 2 seconds.

**Comment by throughthetrees** [comment]
I tried for 5000ms and got the same response.

**Comment by xpeqex** [comment]
Ok, 

We are working to understand what is happening here, but we need some extra information to understand the issue, please share the remaining requested info:

- Could you please share with us the complete output on the Python side?
- What device are you using as a client/slave?
- Is there constant traffic on the bus? A server-client (master-slave) communication, other than the Flipper messages?

**Comment by throughthetrees** [comment]
> What device are you using as a client/slave?

So I'm still learning about this device, but I'm trying to [read from a CWT-CO2-5K-EXPS-D CO2 sensor](https://store.comwintop.com/products/carbon-dioxide-sensor-agricultural-greenhouse-high-precision-industrial-rs485-co2-transmitter).  It came with almost 0 documentation, and almost everything I've found is from guess.

This is what I found on a review of the device that some one else had figured out:

![Screenshot from 2024-07-26 10-16-25](https://github.com/user-attachments/assets/494ec044-891a-46e6-a005-86e264ee97ce)

But afaik, there is no official documentation. I've just been guessing at registers to read from.

Here is the device:

![b40dce48202c326e9e6716f3ee13b1641331a4a0-1](https://github.com/user-attachments/assets/60fd29a9-75f7-4ede-9137-4b552a7775b4)

label view:

![35ad12aca4192d9626131e9ccf2b171a69e30c8d-1](https://github.com/user-attachments/assets/315b7dc5-554b-4b57-8f75-4af7c9e1a9aa)

I am using the electronic cats CatSink variable power supply to get about 2 amp at 12 volt.

![38742d3ed974ad249bdca51011f0425c461a7757-1](https://github.com/user-attachments/assets/c79da8c8-72fa-4923-934e-842a8d4d0809)

and I am bridging to a USB to RS-485 dongle.


So I think the client/ slave answer is just the one device? I only have this device on the modbus and i have to trigger it from either python or flipper to get a reading. I am new to RS-485 and so forgive me if I don't know the terms.

> Could you please share with us the complete output on the Python side?

This is the function I'm running in python:

```
from pymodbus.client import ModbusSerialClient

# Setup the Modbus Serial client
client = ModbusSerialClient(
    method='rtu',
    port='/dev/ttyUSB1',          # Adjust as per your port
    baudrate=4800,                # Set baud rate
    parity='N',                   # Set parity (None, Even, Odd)
    stopbits=1,                   # Number of stop bits
    bytesize=8,                   # Number of data bits
    timeout=.1                     # Timeout in seconds
)

# Connect to the device
connection = client.connect()
if connection:
    # Assuming you want to read holding registers
    response = client.read_holding_registers(address=0x01,count = 16,slave = 1)  # Adjust parameters as needed
    if not response.isError():
        print("Data:", response.registers)
    else:
        print("Error:", response)
    client.close()  # Close the connection
else:
    print("Failed to connect to the device")
    client.close()  # Close the connection
```
and the response I get back is:
```
Data: [0, 452, 0, 0, 0, 0, 0, 0, 3072, 512, 0, 0, 0, 0, 0, 0]
```

The 452 is the ppm CO2, the 3072 is the temperature, and 512 is the humidity (I think).

> Is there constant traffic on the bus? A server-client (master-slave) communication, other than the Flipper messages?

No. Its only traffic if I send request from Python or from flipper.  There are no other devices or commands sending or receiving on the circuit.

And with the help you've already given, it seems like now, yes I can get a signal using manual send!!

So that is good. But it seems like it doesn't parse correctly. You said it seems to be reading in HUB mode?

**Comment by Otrebor671** [comment]
Hi @throughthetrees,
Could you assist us with a test? It would be very helpful in detecting the issue.

We need you to connect the Flipper, the sensor, and the USB to RS485 to your Modbus bus. Then activate the SETTINGS->SAVE LOGS option on the Flipper and set it to SNIFER mode. Now, send the command from Python and wait for the sensor's response.

Once this process is completed, please extract the generated LOG file and send it to us so we can review it for clues to solve the problem.

Please use the latest version of the application, which you can find in the latest release: https://github.com/ElectronicCats/flipper-rs485modbus/releases/tag/v1.1.3.1.

**Comment by throughthetrees** [comment]
Pulling the latest.. will edit when have data.

**Comment by throughthetrees** [comment]
Hey!!

Manual send now works!!!

**Comment by throughthetrees** [comment]
Thank you so much for getting this fixed. Going to begin experimenting now. Also, I would like to learn to contribute, because getting this working was only step one. I want to learn how to add a feature to this library where I continuously make some request (like the above), maybe ever second, and the flipper logs it.

For another time then.

Here is Ka'ala saying thank you.

![dd4d4b281cf582c390c3b4b6c8b93f40007dfb46-1](https://github.com/user-attachments/assets/89085084-819b-4d6e-afbc-9dd5c6515a3c)

**Comment by xpeqex** [comment]
<p align=right>
<img src="https://github.com/ElectronicCats/flipper-shields/assets/107638696/f412a4ae-147b-426b-8d6e-aa41182e0721" width=250>
</p>
